### PR TITLE
Update dependencies and linting + testing setup.

### PR DESCRIPTION
**Pull request by danieldunderfelt**
This PR includes updated linting and testing setup from a project that uses this boilerplate.

### CODE REVIEW

File: .lintstagedrc [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "*.ts?(x)": [
-    "prettier-tslint fix \"src/**/*.ts?(x)\"",
+    "yarn run codestyle",
```
Comment (danieldunderfelt) [inline .lintstagedrc:3]: Unified command to run tslint + Prettier and fix all codestyle issues.

File: config/polyfills.js [hunk 2]
```diff
@@ -11,6 +11,13 @@ if (typeof Promise === 'undefined') {
 // fetch() polyfill for making API calls.
 require('whatwg-fetch');
 
+// Map and Set polyfills
+require('core-js/es6/map');
+require('core-js/es6/set');
```
Comment (danieldunderfelt) [inline config/polyfills.js]: React recommends adding these polyfills.

File: config/polyfills.js [hunk 3]
```diff
@@ -11,6 +11,13 @@ if (typeof Promise === 'undefined') {
...
+
+// RequestAnimationFrame polyfill
+require('raf/polyfill')
```
Comment (danieldunderfelt) [inline config/polyfills.js]: React will actually notify you with a console message if you don't polyfill raf.

File: package.json [hunk 4]
```diff
@@ -1,29 +1,18 @@
 {
```
Comment (danieldunderfelt) [inline package.json:1]: I can't comment on a file in Github (boo!), but I moved all development dependencies to the correct place.

File: package.json [hunk 5]
```diff
@@ -34,84 +23,83 @@
...
   "jest": {
-    "mapCoverage": true,
+    "testEnvironment": "jsdom",
     "collectCoverageFrom": [
       "src/**/*.{js,jsx,ts,tsx}"
     ],
     "setupFiles": [
-      "<rootDir>/config/polyfills.js"
+      "<rootDir>/config/polyfills.js",
+      "<rootDir>/scripts/test.js"
```
Comment (danieldunderfelt) [inline package.json:95]: Jest setup is not done the correct way, instead of running Jest from this file.

File: src/hello/components/Hello.tsx [hunk 6]
```diff
@@ -3,8 +3,8 @@ import { observer } from 'mobx-react'
 import helloStore from '../store'
 import * as css from './Hello.css'
 
-interface Props {}
-interface State {}
+type Props = {}
+type State = {}
```
Comment (danieldunderfelt) [inline src/hello/components/Hello.tsx:7]: Typescript doesn't seem to like empty interfaces, and we decided to use `type` for these if they don't need to extend/be extended.

File: tsconfig.json [hunk 7]
```diff
@@ -23,9 +23,13 @@
...
     "scripts",
+    "config",
     "acceptance-tests",
     "webpack",
     "jest",
-    "src/setupTests.ts"
+    "src/setupTests.ts",
+    "**/*.test.ts",
+    "**/*.test.tsx",
+    "__mocks__"
```
Comment (danieldunderfelt) [inline tsconfig.json:36]: Tslint will complain about these files if they're not excluded from Typescript.

File: tslint.json [hunk 8]
```diff
@@ -1,97 +1,40 @@
...
+    "no-debugger": true,
+    "no-duplicate-variable": true,
+    "no-eval": true,
+    "curly": true,
+    "no-unused-expression": false,
+    "jsx-no-lambda": false,
+    "jsx-no-multiline-js": false,
+    "jsx-boolean-value": false,
+    "no-shadowed-variable": true,
+    "prettier": true
```
Comment (danieldunderfelt) [inline tslint.json:38]: Prettier now runs as a tslint rule, and all codestyle rules are turned off in tslint. Although the linting does seem to take a bit longer, it's more robust this way.

File: typings.d.ts [hunk 9]
```diff
@@ -0,0 +1,8 @@
+declare module '*.css' {
+  interface ClassNames {
+    [className: string]: string
+  }
+
+  const classNames: ClassNames
+  export = classNames
+}
```
Comment (danieldunderfelt) [inline typings.d.ts:8]: Fallback typing for CSS files. Should make development a bit smoother.

**Comment by danieldunderfelt** [comment]
Also updates all dependencies, including Webpack to v4. Migration is done based on this PR to CRA: https://github.com/facebook/create-react-app/compare/next...andriijas:webpack4?diff=unified&name=webpack4#diff-3b55301108483c3bb5289672dd9caa59

**Comment by danieldunderfelt** [comment]
For some reason the Travis build fails on the testing stage with the complaint that there are no tests. This is false. The same command that Travis runs (yarn run ci) works fine locally. What could be the problem?

**Comment by danieldunderfelt** [comment]
Bug found: Jest needs `<rootDir>/` added to ignore config.

**Comment by cniska** [comment]
Awesome thanks!

**Pull request merged**